Add i.MX RT LCD display controller support

MaureenHelm opened the pull request:
Adds a new mcux shim driver for the eLCDIF display controller, and enables it on rt1050 and rt1060 boards. Updates the lvgl sample application to support these boards.

Tested on rt1050 and rt1060.

~~@ioannisg @agansari I cheated a little to configure the mpu for the frame buffer, and just made all ram write through instead of write back. We need a way to partition the ram such that the only the frame buffer is write through (or noncacheable). The ethernet driver needs this too.~~

@galak Please have a look at the dts.

MaureenHelm committed:
dts: Add bindings for the imx lcd display controller

Adds device tree bindings for the imx enhanced LCD interface (eLCDIF)
controller. The compatible is reused from linux.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
drivers: display: Introduce mcux elcdif shim driver

Adds a shim layer around the mcux elcdif driver to adapt it to the
zephyr display interface. Although the hardware and underlying mcux sdk
driver can support additional configurations, some shortcuts are
currently made in the shim that force a given pixel format, lcd data
bus width, and signal polarity. This works with the rocktech lcd module
used on imx rt boards, but will need to be updated for other display
panels.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
soc: nxp_imx: Enable mcux elcdif driver and clocks

Enables the mcux elcdif shim driver and clocks on imx rt socs when the
display driver interface is enabled.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: mimxrt1050_evk: Enable mcux elcdif driver and pins

Enables the mcux elcdif shim driver and pin muxes on the mimxrt1050_evk
board.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
drivers: display: Add RGB 565 pixel format to display api

Adds a new entry in the pixel format enum for RGB 565. This will be used
for the lcd display on the mimxrt1050_evk and mimxrt1060_evk boards.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
samples: gui: Fix lvgl sample to use configured display device name

Fixes the lvgl sample to use the same display device name as the lvgl
library rather than a hardcoded string.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
samples: gui: Add support for imx rt boards to lvgl sample

Adds board configurations to support the mimxrt1050_evk and
mimxrt1060_evk boards in the lvgl sample.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: mimxrt1060_evk: Enable mcux elcdif driver and pins

Enables the mcux elcdif shim driver and pin muxes on the mimxrt1060_evk
board.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

hakehuang reviewed (approved):
test pass in rt1050 evk boards

ioannisg commented:
>@ioannisg @agansari I cheated a little to configure the mpu for the frame buffer, and just made all ram write through instead of write back. We need a way to partition the ram such that the only the frame buffer is write through (or noncacheable). The ethernet driver needs this too.

@MaureenHelm , I have absolutely no objection to having a part of SRAM being write-through, if needed so. The way you have solved it so far, does the following:
- enables _all_ SRAM as write-through, instead of write back, but:
- after we define certain regions _on top_, e.g. Thread Stacks, memory domains, etc., this setting is reverted for those specific regions (I suppose this is OK).

I agree that what we actually need may be _just_ a specific area of SRAM being write-through. This is in accordance with what we have already done, for non-cacheable SRAM part (see #11199, and check with @aurel32 , too). Could we, instead, do something similar for write-through SRAM?

ioannisg reviewed (commented):
Added a comment about how to better provide support for write-through SRAM regions.

aurel32 commented:
> @MaureenHelm , I have absolutely no objection to having a part of SRAM being write-through, if needed so. The way you have solved it so far, does the following:
> 
>     * enables _all_ SRAM as write-through, instead of write back, but:
> 
>     * after we define certain regions _on top_, e.g. Thread Stacks, memory domains, etc., this setting is reverted for those specific regions (I suppose this is OK).

I am not sure it is fine to do that. If the same region is accessed as write-back and then as write-through this causes data loss if the cache is not properly cleaned before the switch. This caused issue in the past (see commit 1397269335). Before doing such a change, we should make sure that the region is defined as write-through in all contexts.
 
> I agree that what we actually need may be _just_ a specific area of SRAM being write-through. This is in accordance with what we have already done, for non-cacheable SRAM part (see #11199, and check with @aurel32 , too). Could we, instead, do something similar for write-through SRAM?

If the region is mostly written, write-through or non-cacheable should not make a big performance difference. The non-cacheable regions are not strictly ordered, so it means that the performances are still correct and that you need a data barrier after the writes. Alternatively if the region is properly aligned, explicit cache clean and invalidate operations might also work.

The non-cached memory regions is not supported with the NXP MPU, but now that I have my IMXRT1020 board working, I can try to have a look.

agansari commented:
@MaureenHelm regarding write-through, the Ethernet driver can work with no cache descriptors and wb cache, as i have enabled SDK's ethernet driver internal cache management with FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL.

Setting all RAM as no cache obviously has a performance impact on networking.

I've made this premature pull request so you can checkout no cache: #12465

agansari commented:
@MaureenHelm 
We could define a write through region similar to CONFIG_NOCACHE_MEMORY that works on M7 cores.

But I prefer the a cache flushing solution on write back memory. this task generates heavy memory traffic (memcpy + lcd controller's dma). Cache flushing (clean) can be uses with fsl_cache.h or core_m7.h

MaureenHelm commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -0,0 +1,271 @@
+
+LOG_MODULE_REGISTER(display_mcux_elcdif, CONFIG_DISPLAY_LOG_LEVEL);
+
+K_MEM_POOL_DEFINE(mcux_elcdif_pool,
```
Is there a reason syscalls haven't been implemented?

vanwinkeljan replied:
No until now all display drivers where either using SPI or I2C, so there was no need for syscalls

MaureenHelm commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -0,0 +1,271 @@
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	memcpy(data->fb[write_idx].data, data->fb[read_idx].data,
```
Agree. It would also allow multiple writes per frame, which we currently can't do.

MaureenHelm commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -0,0 +1,271 @@
+			    const struct display_buffer_descriptor *desc,
+			    void *buf)
+{
+	return -ENOTSUP;
```
Because we don't have any consumers of this API. And I was lazy.

MaureenHelm commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -0,0 +1,271 @@
+
+static int mcux_elcdif_display_blanking_off(const struct device *dev)
+{
+	return 0;
```
Agreed

MaureenHelm commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -0,0 +1,271 @@
+					const enum display_pixel_format
+					pixel_format)
+{
+	if (pixel_format == PIXEL_FORMAT_RGB_565) {
```
This is a limitation of the display. The controller can support other formats, but I don't have a way to test any of them right now.

This should most likely be handled in the device tree, but I'm not sure exactly how. Need some guidance from @galak.

vanwinkeljan replied:
Also interested in how to solve this with DTS, for the ILI display driver the display is selected through Kconfig for now

MaureenHelm commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -0,0 +1,271 @@
+	struct mcux_elcdif_data *data = dev->driver_data;
+	int i;
+
+	u32_t polarity_flags = kELCDIF_DataEnableActiveHigh |
```
Agree. I think this should also be in the device tree.

MaureenHelm commented on soc/arm/nxp_imx/rt/soc.c:
```diff
@@ -48,6 +48,15 @@ const clock_enet_pll_config_t ethPllConfig = {
 
+#ifdef CONFIG_INIT_VIDEO_PLL
+const clock_video_pll_config_t videoPllConfig = {
+	.loopDivider = 31,
```
We should, but I'd like to wait to do that since there are other clock dividers that need it too.

MaureenHelm commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -0,0 +1,271 @@
+
+	LOG_DBG("W=%d, H=%d, @%d,%d", desc->width, desc->height, x, y);
+
+	k_sem_take(&data->sem, K_FOREVER);
```
The intent of the semaphore is to throttle writes such that we don't swap the buffers more than once per frame. If a write hasn't been called in a while (>1 frame), then the semaphore will be immediately available.

dbkinder reviewed (approved):
doc changes LGTM.

agansari reviewed (approved):
> Removes the write-through changes and adds cache cleaning to the display driver

+1

galak commented:
Looking at how panel dts are supported in linux for this simple panel I think you can do something like:

[ My `remote-endpoint' prop's need fixing ]

```
        panel {
                compatible = "rocktech,rk070er9427";
                backlight = <&backlight_lcd>;

                port {
                        lcd_panel_in: endpoint {
                                remote-endpoint = <&lcd_display_out>;
                        };
                };
        };
        
        
        lcdif1: display-controller@2220000 {
                compatible = "fsl,imx6sx-lcdif", "fsl,imx28-lcdif";
                reg = <0x02220000 0x4000>;
                interrupts = <GIC_SPI 5 IRQ_TYPE_LEVEL_HIGH>;
                clocks = <&clks IMX6SX_CLK_LCDIF1_PIX>,
                         <&clks IMX6SX_CLK_LCDIF_APB>,
                         <&clks IMX6SX_CLK_DISPLAY_AXI>;
                clock-names = "pix", "axi", "disp_axi";

                port {
                        parallel_out: endpoint {
                                remote-endpoint = <&panel_in_parallel>;
                        };
                };
        };
```

jhqian reviewed (commented):
Hi @MaureenHelm, as RT1064 has been merged, will you add that as well?

MaureenHelm commented:
> Hi @MaureenHelm, as RT1064 has been merged, will you add that as well?

@jhqian: Yes. Can you help me with testing? I don't have enough displays at my desk to dedicate one for each RT board, and the connectors are too fragile to frequently move the display from board to board.

> Looking at how panel dts are supported in linux for this simple panel I think you can do something like:

@galak: Where should this stuff go? I realize it's deprecated in the fsl linux bindings, but it seems like a step in the wrong direction to put these parameters in Kconfig.
```
	bits-per-pixel = <16>;
	bus-width = <16>;
	hactive = <480>;
	vactive = <272>;
	hback-porch = <8>;
	hfront-porch = <4>;
	hsync-len = <41>;
	vback-porch = <2>;
	vfront-porch = <4>;
	vsync-len = <10>;
```

MaureenHelm committed:
dts: boards: Add bindings and nodes for the rocktech lcd module

Adds device tree bindings and nodes for the rocktech lcd module used on
imx rt boards. The use of port, endpoint, and remote-endpoint properties
follow what is currently done in linux, but they are not yet used in
zephyr because some additional plumbing is needed in the
extract_dts_includes.py script.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm commented:
Rebased to latest master and reworked the dts. @galak, please have another look.

galak reviewed (dismissed):
Rather than putting all this in Kconfig:

```
	.hactive = CONFIG_MCUX_ELCDIF_HACTIVE,
	.vactive = CONFIG_MCUX_ELCDIF_VACTIVE,
	.hback_porch = CONFIG_MCUX_ELCDIF_HBACK_PORCH,
	.hfront_porch = CONFIG_MCUX_ELCDIF_HFRONT_PORCH,
	.hsync_len = CONFIG_MCUX_ELCDIF_HSYNC_LEN,
	.vback_porch = CONFIG_MCUX_ELCDIF_VBACK_PORCH,
	.vfront_porch = CONFIG_MCUX_ELCDIF_VFRONT_PORCH,
	.vsync_len = CONFIG_MCUX_ELCDIF_VSYNC_LEN,
	.bits_per_pixel = CONFIG_MCUX_ELCDIF_BITS_PER_PIXEL,
	.bus_width = CONFIG_MCUX_ELCDIF_BUS_WIDTH,
```

Can we just have it be the display type and have all the param's defined based on a display single Kconfig?

MaureenHelm commented:
> Can we just have it be the display type and have all the param's defined based on a display single Kconfig?

Sure, done.

MaureenHelm commented:
@vanwinkeljan 	Please revisit. I think I've addressed most of your issues except for userspace and flushing, which I'd like to postpone to future PRs.

vanwinkeljan commented on drivers/display/display_mcux_elcdif.c:
```diff
@@ -0,0 +1,286 @@
+		.vsw = config->vsync_len,
+		.vfp = config->vfront_porch,
+		.vbp = config->vback_porch,
+		.polarityFlags = polarity_flags,
```
Could polarity, pixelFomrat and databus not been taken from the config struct, hey also look like display panel specific parameters?

MaureenHelm replied:
Yes, done

vanwinkeljan reviewed (changes requested):
LGTM except one small remark regarding the display specific parameters
